pkg/distributor/ratestore.go

```diff
@@ -98,11 +106,28 @@ func (s *rateStore) updateAllRates(ctx context.Context) error {
 	streamRates := s.getRates(ctx, clients)
 	rates := s.aggregateByShard(streamRates)
 
-	s.metrics.streamCount.Set(float64(len(rates)))
-
 	s.rateLock.Lock()
 	defer s.rateLock.Unlock()
-	s.rates = rates
+
+	for stream, rate := range rates {
+		s.rates[stream] = expiringRate{
+			rate:      rate,
+			createdAt: time.Now(),
+		}
```

Comment:
i think you can check for `time.Since(rate.createdAt) > s.rateKeepAlive` and update maxRate here and get rid of the next iteration. the bigO complexity is the same but I feel like it should have better performance since you would avoid calling `delete`

Reply:
I think we need to do both: There is no guarantee that both maps contain the same entries so first we merge them and then we iterate over the whole map to make sure we don't miss anything.

Reply:
got it, at first I thought that all entries on `s.rates` would be on the initial `rates`.